File: port/kotlin/src/main/kotlin/muonledger/Journal.kt
package muonledger

/**
 * Central container for all financial data in a Ledger session.
 *
 * Mirrors ledger's `journal_t` type. The journal owns the account tree
 * (rooted at [root]), all parsed transactions, and account aliases.
 */
class Journal {

    /** The root account node (invisible, depth 0). */
    val root: Account = Account()

    /** All regular transactions, in parse order. */
    val transactions: MutableList<Xact> = mutableListOf()

    /** Account aliases: short name -> full path. */
    val aliases: MutableMap<String, String> = mutableMapOf()

    /**
     * Add a transaction to the journal after finalizing it.
     *
     * Calls [Xact.finalize] to infer missing amounts and verify
     * double-entry balance, then appends the transaction.
     *
     * @throws BalanceError if the transaction does not balance.
     */
    fun addTransaction(xact: Xact) {
        xact.finalize()
        transactions.add(xact)
    }

    /**
     * Look up or create an account by colon-separated path.
     *
     * If an alias exists for the given name, it is resolved first.
     */
    fun findOrCreateAccount(name: String): Account {
        val resolved = aliases[name] ?: name
        return root.findAccount(resolved)!!
    }

    /** Number of transactions in the journal. */
    val size: Int get() = transactions.size

    override fun toString(): String {
        return "Journal(transactions=${transactions.size}, accounts=${root.flatten().size})"
    }
}

File: port/swift/Sources/MuonLedger/Journal.swift
/// Central container for all financial data in a ledger session.
///
/// Ported from ledger's `journal_t`. The Journal owns the account tree
/// (rooted at a master account), all parsed transactions, and account aliases.
///
/// There is exactly one journal per session. Parsing populates it and
/// the reporting engine reads from it.

import Foundation

// MARK: - Journal

/// The central container for all financial data.
public final class Journal {
    /// Root account (invisible, empty name).
    public var master: Account

    /// All regular transactions in parse order.
    public var transactions: [Xact]

    /// Account name aliases.
    public var aliases: [String: String]

    /// Create a new empty journal.
    public init() {
        self.master = Account()
        self.transactions = []
        self.aliases = [:]
    }

    // MARK: Transaction management

    /// Add a transaction to the journal after finalizing it.
    ///
    /// Calls `finalize()` to infer missing amounts and verify balance.
    /// Returns `true` if the transaction was successfully added, `false`
    /// if finalization indicated it should be skipped.
    ///
    /// - Throws: `TransactionError` if the transaction does not balance.
    @discardableResult
    public func addTransaction(_ xact: Xact) throws -> Bool {
        guard try xact.finalize() else { return false }
        transactions.append(xact)
        return true
    }

    // MARK: Account helpers

    /// Look up or create an account by colon-separated path.
    ///
    /// Resolves aliases before lookup. For example, if an alias maps
    /// "Food" to "Expenses:Food", then `findOrCreateAccount("Food")`
    /// returns the `Expenses:Food` account.
    public func findOrCreateAccount(_ name: String) -> Account {
        let resolvedName = aliases[name] ?? name
        return master.findAccount(resolvedName, autoCreate: true)!
    }

    // MARK: Container protocol

    /// Number of transactions in the journal.
    public var count: Int {
        transactions.count
    }

    // MARK: Reset

    /// Reset the journal to its initial (empty) state.
    public func clear() {
        master = Account()
        transactions.removeAll()
        aliases.removeAll()
    }
}
